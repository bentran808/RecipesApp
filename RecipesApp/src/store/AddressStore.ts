import { toJS } from 'mobx';
import { types, destroy, SnapshotOut, getRoot, applySnapshot } from 'mobx-state-tree';
import { RootStore } from 'store/store';

const AddressEntry = types
  .model('AddressEntry', {
    id: types.optional(types.identifierNumber, 0),
    type: types.optional(types.string, ''),
    address: types.optional(types.string, '')
  })
  .actions((self) => ({
    remove() {
      getRoot<typeof RootStore>(self).address.removeItem(self);
    },
    edit(type: string, address: string) {
      self.type = type;
      self.address = address;
    }
  }));

export type AddressModel = SnapshotOut<typeof AddressEntry>;

const AddressStore = types
  .model('', {
    items: types.array(AddressEntry),
    itemUsing: types.optional(AddressEntry, {})
  })
  .volatile(() => ({
    address: ''
  }))
  .actions((self) => ({
    setAddress(address: string) {
      self.address = address;
    },
    addNewItem(type: string, address: string) {
      const id = self.items.reduce((maxId, item) => Math.max(item.id, maxId), -1) + 1;
      self.items.push({ id, type, address });
    },
    removeItem(item: AddressModel) {
      destroy(item);
    },
    existAddress(address?: AddressModel) {
      return self.items.find((item) => item.id === address?.id);
    },
    setItemUsing(item: AddressModel) {
      applySnapshot(self, {
        ...self,
        itemUsing: item
      });
    }
  }))
  .views((self) => ({
    get addresses() {
      return toJS(self.items);
    }
  }));

export default AddressStore;
