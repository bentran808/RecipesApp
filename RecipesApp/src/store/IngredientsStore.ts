import { recipesApi } from 'api';
import { toJS } from 'mobx';
import { flow, SnapshotOut, types } from 'mobx-state-tree';

export const IngredientsEntry = types.model('IngredientsEntry', {
  id: types.number,
  name: types.string,
  photo_url: types.string
});

export type IngredientModel = SnapshotOut<typeof IngredientsEntry>;

const IngredientsStore = types
  .model('IngredientsStore', {
    lists: types.optional(types.array(IngredientsEntry), []),
    state: types.optional(types.enumeration('State', ['pending', 'done', 'error']), 'done')
  })
  .views((self) => ({
    get ingredients() {
      return self.lists;
    },
    get ingredientsJS() {
      return toJS(this.ingredients);
    }
  }))
  .actions((self) => ({
    fetchIngredients: flow(function* fetchIngredients() {
      self.state = 'pending';
      try {
        const response = yield recipesApi.fetchIngredientsRequest();
        self.lists = response.data;
        self.state = 'done';
      } catch (error) {
        self.state = 'error';
      }
    })
  }));

export default IngredientsStore;
