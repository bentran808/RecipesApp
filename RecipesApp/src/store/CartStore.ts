import { toJS } from 'mobx';
import { applySnapshot, destroy, getRoot, SnapshotOut, types } from 'mobx-state-tree';
import { RecipeModel, RecipesEntry } from 'store/RecipesStore';
import { RootStore } from 'store/store';

export const CartEntry = types
  .model('CartEntry', {
    item: RecipesEntry,
    quantity: types.optional(types.number, 1),
    inCart: types.optional(types.boolean, true)
  })
  .actions((self) => ({
    remove() {
      getRoot<typeof RootStore>(self).cart.removeItem(self);
    },
    increase(quantity?: number) {
      self.quantity = quantity ? self.quantity + quantity : ++self.quantity;
    },
    decrease() {
      self.quantity === 1 ? this.remove() : --self.quantity;
    },
    toggle() {
      self.inCart = !self.inCart;
    }
  }));

const BillEntry = types.model('BillEntry', {
  type: types.optional(types.string, ''),
  name: types.optional(types.string, ''),
  code: types.optional(types.string, ''),
  price: types.optional(types.number, 0)
});

const OrderedEntry = types.model('OrderedEntry', {
  items: types.array(
    types.model({
      name: types.optional(types.string, ''),
      quantity: types.optional(types.number, 0)
    })
  ),
  total: types.number,
  createdAt: types.string
});

export type CartModel = SnapshotOut<typeof CartEntry>;
export type BillModel = SnapshotOut<typeof BillEntry>;
export type OrderedModel = SnapshotOut<typeof OrderedEntry>;

const CartStore = types
  .model('CartStore', {
    items: types.array(CartEntry),
    billItems: types.array(BillEntry),
    orderedItems: types.array(OrderedEntry)
  })
  .views((self) => ({
    get recipes() {
      return toJS(self.items);
    },
    get recipesInCart() {
      return this.recipes.filter((item) => item.inCart);
    },
    get inCartCount() {
      return this.recipesInCart.length;
    },
    get total() {
      return this.recipesInCart.reduce((sum, item) => sum + item.quantity * 20, 0);
    },
    get billDetails() {
      return toJS(self.billItems);
    },
    get toPay() {
      return this.billDetails.reduce(
        (sum, item) => (item.type === 'discount' ? sum - item.price : sum + item.price),
        0
      );
    },
    get orderedList() {
      return toJS(self.orderedItems);
    }
  }))
  .volatile(() => ({
    discountInput: '',
    invalidCoupon: false
  }))
  .actions((self) => ({
    addToCart(item: RecipeModel, quantity?: number) {
      const itemInCart = this.hasInCart(item);
      itemInCart ? itemInCart.increase(quantity) : self.items.push({ item, quantity });
    },
    removeItem(item: CartModel) {
      destroy(item);
    },
    hasInCart(recipe: RecipeModel) {
      return self.items.find((item) => item.item.id === recipe.id);
    },
    setDiscountInput(text: string) {
      if (!text) {
        self.invalidCoupon = false;
      }
      self.discountInput = text;
    },
    applyCoupon(item: BillModel) {
      const existItem = self.billItems.findIndex((bill) => bill.type === 'discount');
      if (existItem > -1) {
        self.billItems.splice(existItem, 1, item);
      } else {
        self.billItems.push(item);
      }
    },
    payment(order: OrderedModel) {
      self.orderedItems.unshift(order);
      applySnapshot(self, {
        ...self,
        items: [],
        billItems: []
      });
    },
    isInvalidCoupon(status: boolean) {
      self.invalidCoupon = status;
    }
  }));

export default CartStore;
