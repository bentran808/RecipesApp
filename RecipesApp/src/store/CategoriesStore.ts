import { recipesApi } from 'api';
import { toJS } from 'mobx';
import { flow, SnapshotOut, types } from 'mobx-state-tree';
import { transformCategories } from 'utils';
import { RecipesEntry } from './RecipesStore';

export const CategoriesEntry = types.model('CategoriesEntry', {
  id: types.number,
  name: types.string,
  photo_url: types.string,
  recipes: types.maybe(types.array(RecipesEntry))
});

export type CategoryModel = SnapshotOut<typeof CategoriesEntry>;

const CategoriesStore = types
  .model('CategoriesStore', {
    lists: types.optional(types.array(CategoriesEntry), []),
    state: types.optional(types.enumeration('State', ['pending', 'done', 'error']), 'done')
  })
  .views((self) => ({
    get categories() {
      return self.lists;
    },
    get categoriesJS() {
      return transformCategories(toJS(this.categories));
    }
  }))
  .actions((self) => ({
    fetchCategories: flow(function* fetchCategories() {
      self.state = 'pending';
      try {
        const response = yield recipesApi.fetchCategoriesRequest();
        self.lists = response.data;
        self.state = 'done';
      } catch (error) {
        console.log(error);
        self.state = 'error';
      }
    })
  }));

export default CategoriesStore;
