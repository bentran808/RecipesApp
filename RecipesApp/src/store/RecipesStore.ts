import { recipesApi } from 'api';
import { toJS } from 'mobx';
import { applySnapshot, flow, SnapshotOut, types } from 'mobx-state-tree';
import { CategoryModel } from 'store/CategoriesStore';

export const RecipesEntry = types.model('RecipesEntry', {
  id: types.number,
  categoryId: types.number,
  title: types.string,
  photo_url: types.string,
  photosArray: types.array(types.string),
  time: types.string,
  ingredients: types.array(types.array(types.union(types.number, types.string))),
  description: types.string,
  category: types.maybe(
    types.model({
      id: types.number,
      name: types.string,
      photo_url: types.string
    })
  )
});

export type RecipeModel = SnapshotOut<typeof RecipesEntry>;

const RecipesStore = types
  .model('RecipesStore', {
    items: types.optional(types.array(RecipesEntry), []),
    state: types.optional(types.enumeration('State', ['pending', 'done', 'error']), 'done'),
    currentPage: types.optional(types.number, 1),
    isEndList: types.optional(types.boolean, false)
  })
  .views((self) => ({
    get recipesJS() {
      return toJS(self.items);
    }
  }))
  .volatile(() => ({
    keyword: ''
  }))
  .actions((self) => ({
    fetchRecipes: flow(function* fetchRecipes(page?: number) {
      if (!self.isEndList) {
        self.state = 'pending';
        try {
          const response = yield recipesApi.fetchRecipesRequest(page);
          if (response.data.length) {
            self.currentPage++;
            applySnapshot(self, {
              ...self,
              items: !page || page === 1 ? response.data : [...self.items, ...response.data]
            });
          } else {
            self.isEndList = true;
          }
          self.state = 'done';
        } catch (error) {
          self.state = 'error';
        }
      }
    }),
    fetchRecipesByCategoryId: flow(function* fetchRecipesByCategoryId(id: number) {
      self.state = 'pending';
      if (self.isEndList) {
        self.currentPage = 1;
        self.isEndList = false;
      }
      try {
        const response = yield recipesApi.fetchRecipesByCategoryIdRequest(id);
        self.items = response.data;
        self.state = 'done';
      } catch (error) {
        self.state = 'error';
      }
    }),
    searchRecipeName: flow(function* searchRecipeName(text: string) {
      self.state = 'pending';
      if (self.isEndList) {
        self.currentPage = 1;
        self.isEndList = false;
      }
      try {
        const response = yield recipesApi.searchByRecipeNameRequest(text);

        applySnapshot(self, {
          ...self,
          items: [...self.items, ...response.data]
        });
        self.state = 'done';
      } catch (error) {
        self.state = 'error';
      }
    }),
    searchCategoryName: flow(function* searchCategoryName(text: string) {
      self.state = 'pending';
      if (self.isEndList) {
        self.currentPage = 1;
        self.isEndList = false;
      }
      try {
        const response = yield recipesApi.searchByCategoryNameRequest(text);
        const recipesOfCategory = (response.data || [])
          .map((category: CategoryModel) =>
            (category.recipes || []).map((recipe) => ({
              ...recipe,
              category: {
                id: category.id,
                name: category.name,
                photo_url: category.photo_url
              }
            }))
          )
          .flat(1);

        applySnapshot(self, {
          ...self,
          items: [...self.items, ...recipesOfCategory]
        });
        self.state = 'done';
      } catch (error) {
        self.state = 'error';
      }
    }),
    setKeyword(text: string) {
      self.keyword = text;
    },
    setRefreshData() {
      self.currentPage = 1;
      self.isEndList = false;
    },
    setEmptyRecipes() {
      applySnapshot(self, {
        ...self,
        items: []
      });
    }
  }));

export default RecipesStore;
