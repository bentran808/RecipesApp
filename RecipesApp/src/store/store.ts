import { Instance, types } from 'mobx-state-tree';
import RecipesStore from 'store/RecipesStore';
import CategoriesStore from './CategoriesStore';

const RootStore = types.model('RootStore', {
  categories: CategoriesStore,
  recipes: RecipesStore
});

const store = RootStore.create({
    categories: {},
    recipes: {}
});

export type RootStoreModel = Instance<typeof RootStore>;

export default store;
