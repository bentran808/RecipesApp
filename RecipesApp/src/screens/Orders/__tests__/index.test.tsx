import { StoreProvider } from 'context';
import React from 'react';
import renderer from 'react-test-renderer';
import { RootStore } from 'store/store';
import OrdersScreen from '..';

describe('Orders Screen Component', () => {
  const appStore = RootStore.create({
    categories: {},
    recipes: {},
    ingredients: {},
    cart: {
      orderedItems: [
        {
          items: [
            {
              name: 'test',
              quantity: 2
            }
          ],
          total: 40,
          createdAt: ''
        },
        {
          items: [
            {
              name: 'test 2',
              quantity: 1
            }
          ],
          total: 20,
          createdAt: ''
        }
      ]
    },
    address: {}
  });

  test('should render correctly', () => {
    const tree = renderer
      .create(
        <StoreProvider value={appStore}>
          <OrdersScreen />
        </StoreProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
