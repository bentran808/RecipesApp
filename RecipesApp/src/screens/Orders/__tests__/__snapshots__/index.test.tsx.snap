// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orders Screen Component should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      Array [
        Object {
          "createdAt": "",
          "items": Array [
            Object {
              "name": "test",
              "quantity": 2,
            },
          ],
          "total": 40,
        },
        Object {
          "createdAt": "",
          "items": Array [
            Object {
              "name": "test 2",
              "quantity": 1,
            },
          ],
          "total": 20,
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    testID="orderedList"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "paddingBottom": 15,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingHorizontal": 20,
                "paddingVertical": 10,
                "width": 750,
              }
            }
          >
            <View
              style={
                Object {
                  "maxWidth": 200,
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontWeight": "bold",
                  }
                }
              >
                test x 2
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "gray",
                    },
                    Object {
                      "fontSize": 12,
                      "marginTop": 5,
                    },
                  ]
                }
              >
                
              </Text>
            </View>
            <Text>
              $
              40.00
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="reorderBtn"
          >
            <View
              style={
                Array [
                  Object {
                    "borderRadius": 100,
                    "marginHorizontal": 10,
                  },
                  Object {
                    "borderColor": "red",
                    "borderWidth": 1,
                    "paddingHorizontal": 30,
                    "paddingVertical": 5,
                  },
                  Object {
                    "borderRadius": 5,
                    "marginLeft": 20,
                    "marginTop": 10,
                    "width": 130,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "red",
                      "fontSize": 14,
                      "fontWeight": "bold",
                      "textAlign": "center",
                    },
                  ]
                }
              >
                REORDER
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "borderBottomColor": "gray",
              "borderBottomWidth": 1,
              "marginHorizontal": 30,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "paddingBottom": 15,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingHorizontal": 20,
                "paddingVertical": 10,
                "width": 750,
              }
            }
          >
            <View
              style={
                Object {
                  "maxWidth": 200,
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontWeight": "bold",
                  }
                }
              >
                test 2 x 1
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "gray",
                    },
                    Object {
                      "fontSize": 12,
                      "marginTop": 5,
                    },
                  ]
                }
              >
                
              </Text>
            </View>
            <Text>
              $
              20.00
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="reorderBtn"
          >
            <View
              style={
                Array [
                  Object {
                    "borderRadius": 100,
                    "marginHorizontal": 10,
                  },
                  Object {
                    "borderColor": "red",
                    "borderWidth": 1,
                    "paddingHorizontal": 30,
                    "paddingVertical": 5,
                  },
                  Object {
                    "borderRadius": 5,
                    "marginLeft": 20,
                    "marginTop": 10,
                    "width": 130,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "red",
                      "fontSize": 14,
                      "fontWeight": "bold",
                      "textAlign": "center",
                    },
                  ]
                }
              >
                REORDER
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
